### PR TITLE
Allow public use of DBEntry

### PR DESCRIPTION
**Pull request by thomascobb**

### CODE REVIEW

File: epicsdbbuilder/dbd.py [hunk 1]
```diff
@@ -101,6 +96,35 @@ def ValidFieldValue(self, name, value):
 # the DBD entries are accumulated into a single large database.
 _db = ctypes.c_void_p()
 
+
+class DBEntry:
+    def __init__(self, entry=None):
+        if entry is None:
+            assert _db, "LoadDdbFile not called yet"
```
Comment (Araneidae) [inline epicsdbbuilder/dbd.py:103]: Do this unconditionally

File: epicsdbbuilder/dbd.py [hunk 2]
```diff
@@ -101,6 +96,35 @@ def ValidFieldValue(self, name, value):
...
+        status = mydbstatic.dbFirstField(self, dct_only)
+        while status == 0:
+            yield mydbstatic.dbGetFieldName(self)
+            status = mydbstatic.dbNextField(self, dct_only)
+
+    def copy(self):
+        entry = mydbstatic.dbCopyEntry(self)
+        return DBEntry(entry)
+
+    def free(self):
```
Comment (Araneidae) [inline epicsdbbuilder/dbd.py:123]: Think we can use `__del__` instead.

File: epicsdbbuilder/dbd.py [hunk 3]
```diff
@@ -101,6 +96,35 @@ def ValidFieldValue(self, name, value):
...
+            status = mydbstatic.dbNextRecordType(self)
+
+    def iterate_fields(self, dct_only=0):
+        status = mydbstatic.dbFirstField(self, dct_only)
+        while status == 0:
+            yield mydbstatic.dbGetFieldName(self)
+            status = mydbstatic.dbNextField(self, dct_only)
+
+    def copy(self):
+        entry = mydbstatic.dbCopyEntry(self)
```
Comment (Araneidae) [inline epicsdbbuilder/dbd.py:120]: Cleaner to call `dbCopyEntry` in the constructor?

File: epicsdbbuilder/dbd.py [hunk 4]
```diff
@@ -122,18 +146,14 @@ def LoadDbdFile(dbdfile, on_use = None):
...
-        if not hasattr(records, recordType):
+    entry = DBEntry()
+    for record_type in entry.iterate_records():
+        if not hasattr(records, record_type):
             validate = ValidateDbField(entry)
-            records._PublishRecordType(on_use, recordType, validate)
-        status = mydbstatic.dbNextRecordType(entry)
-    mydbstatic.dbFreeEntry(entry)
+            records._PublishRecordType(on_use, record_type, validate)
+    entry.free()
```
Comment (Araneidae) [inline epicsdbbuilder/dbd.py:156]: Should be able drop this now.

File: epicsdbbuilder/mydbstatic.py [hunk 5]
```diff
@@ -83,6 +83,13 @@ def ImportFunctions(epics_base, host_arch):
             epics_base, library_location, host_arch,
             library_name_format.format('dbStaticHost')))
 
+_libdb = None
+
+
+def GetFunction(name):
```
Comment (Araneidae) [inline epicsdbbuilder/mydbstatic.py:89]: Maybe call this `GetDbFunction`?

File: docs/api.rst [hunk 6]
```diff
@@ -343,3 +343,25 @@ Record Class
...
+
+Advanced usage may require using other functions from the dbCore library to 
+get extra introspection information about records.
+
+..  py:currentmodule:: epicsdbbuilder.mydbstatic
+..  function:: GetDbFunction(name)
+
+    This can be used to get other ``dbCore`` functions for advanced usage.  Note
+    that before use their argtypes will need to be declared in the standard
+    `ctypes` way.
```
Comment (Araneidae) [inline docs/api.rst:359]: Hmmm.  How is the user expected to apply these argtypes?  Could the necessary parameters be passed as optional arguments to `GetDbFunction`?

File: epicsdbbuilder/mydbstatic.py [hunk 7]
```diff
@@ -86,10 +86,15 @@ def ImportFunctions(epics_base, host_arch):
 _libdb = None
 
 
-def GetDbFunction(name):
+def GetDbFunction(name, restype, argtypes, errcheck=None):
```
Comment (Araneidae) [inline epicsdbbuilder/mydbstatic.py:89]: These could be optional: I think `ctypes` has vaguely sensible defaults, and is it possible that argtypes have already been set elsewhere?  I'm not sure what happens if you ask for the same function twice from a library: do you get the same object?  If so, it's quite possible these may have already been set...

**Comment by thomascobb** [comment]
Can I merge this now?

**Comment by Araneidae** [comment]
Go ahead.

**Pull request merged**